src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -54,18 +57,33 @@ public void checkUserInDatastore(User user) {
     // ** NEED TO WRITE TEST FOR BEHAVIORAL TESTING
 
     Key userKey = KeyFactory.createKey("User", user.getUserId());
-    if (datastore.get(userkey) == null) {
-        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-
-        // Create new user entity 
-        Entity userEntity = user.createEntity();
-        Key userKey = userEntity.getKey();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity userEntity;
 
-        // Create User's key and add to object 
-        String userKeyStr = KeyFactory.keyToString(userKey);
-        user.setUserKey(userKeyStr);
+    // Check if user is in datastore
+    try {
+      userEntity = datastore.get(userKey);
+      user.setUserKey(KeyFactory.keyToString(userKey));
+    } catch (EntityNotFoundException e) {
+      userEntity = null;
+    }
 
-        datastore.put(userEntity);
+    if (userEntity == null) {
+        addUserToDatastore(datastore, user);
```

Comment:
If the userId is null then you should call the addUserToDatastore when you check on line 48. That way you don't also make a unnecessary request to datastore.